### DISCUSSION TITLE
What is the difference between Invariant and Equivariant messages?

### DISCUSSION THREAD

**Discussion by pobo95** [ACEsuit/mace#556] [Q&A] [answered]
Hi,
Thank you for your great work and I have some questions for channel.

Using `--hidden_irreps`, we can control the number of messages.

 `--hidden_irreps='256x0e'` for invariant messages and  `--hidden_irreps='256x0e + 256x1o'`  for equivariant messages.

Here is my questions.

1. What exactly does it mean by invariant and equivariant?
Does the invariant messages only use the distance of atoms? (l = 0 for spherical harmonics)

2. Does  `--hidden_irreps` control the number of hidden layers?
If I used `--hidden_irreps='256x0e + 256x1o'` , 
does it mean 256 invarinat messages for the first hidden layer and 256 equivariant messages for the second hidden layer?

Then can I used like this?  `--hidden_irreps='256x1o'` 256 equivariant messages for one hidden layer.

**Comment by ilyes319** [comment, marked as answer]
I recommend you to read our different papers in details to understand these points.

Because MACE constructs many body messages, even an invariant mace layer is not just a function of the distances but of other invariant many body quantities (think angle, dihedral, even if it is not implemented like that).

The hidden irreps controls the number of channels and equivariance of the message. Having it being "128x0e + 128x1o" means construct 128 scalar messages and 128 l=1 vectorial messages. Each layer passes the same message. 

To understand more about equivariance, you can read the e3nn tutorial.

**Reply by nihil39** [reply]
How do you construct the messages starting from node features and distances between nodes?

Are the messages passed the basis function used in ACE (so Radial functions, spherical harmonics and a node embedding)? 
I have read [this](https://arxiv.org/abs/2312.07511) (famous?) paper and I have understood how to raise or lower the order of a tensor with generalized scalar and outer product operations but it is still not clear to me what are the messages passed in MACE when for example you set L max = 2. 

Thank you

I don't know if it would have been better to open a new thread to ask this question.

**Reply by ilyes319** [reply]
@nihil39 Hello, I would recommend you read in detail the MACE paper where this is explained and do our architecture tutorial here: https://colab.research.google.com/drive/1AlfjQETV_jZ0JQnV5M3FGwAM2SGCl2aU

**Reply by nihil39** [reply]
Thank you. I have started to study the paper and do the notebook tutorial. 

Some questions
How do you compute the node features for each element? The number of channels is an hyperparameter which should change the accuracy of the network in a general sense. 

You use `LinearNodeEmbeddingBlock`, and you are doing the product of this vector `[1,0,0] `(one hot encoding of node attributes for hydrogen, for example) with what? 
The results depends on the selected number of channels, is the number of channel the number of input neurons of the MLP?

Thank you and sorry if these questions are too trivial for you.

**Reply by nihil39** [reply]
Other question.

In the following picture l_max = 2 (small l). The default value as far as I have understood. What would happen if L _max (capital L) = 1 for example? Would it mean that you would keep only the m = 0 and m = +1 and -1 components so "cutting"  the -2 and +2 components of the spherical harmonics of order 2?

![image](https://github.com/user-attachments/assets/25d681a8-1e46-47cd-aaeb-7bb3b945fd20)

**Reply by lucasdekam** [reply]
Hey @nihil39 , thanks for your questions in this discussion forum and the link to the 'hitchhiker's guide' paper, I'm also learning about equivariant GNNs and it's very helpful to read :-) 

I think I can answer your question about $L_\mathrm{max}$. $L_\mathrm{max}$ is the tensor order of the message. I think you can roughly think of it as follows. The message is constructed using the ACE basis (which uses spherical harmonics for the radial part); these basis functions have a value of $l$ associated with it. Within e3nn, you can specify a tensor product in terms of what tensor orders $l$ you put in and what tensor orders $l$ you get out. If ell_max $l_\mathrm{max}$=3, we use $l=0,1,2,3$ for the ACE basis as input, and if $L_\mathrm{max}=1$, we want outputs of $l=0,1$. e3nn then figures out which combinations (paths) are allowed (I think using Clebsch-Gordan coefficients). Higher-order outputs are discarded. In the picture you shared, this would mean we discard the entire bottom row.

In the MACE paper, the tensor product is described in steps. First the ACE A-basis functions (with their corresponding l) are combined in a tensor product to get B-basis functions (with corresponding L). Then the message is written as a linear combination of B-basis functions (so the tensor order L).

As a rule of thumb: in equivariant operations, the components of spherical tensors (m=-1, 0, 1...) are never handled individually. You always do things in blocks corresponding to one value of $l$.

Hopefully someone correct me if I'm wrong, because I'm still pretty new to this as well.

**Reply by nihil39** [reply]
Thanks a lot. But all these equivariant networks are related to message passing, right? So the features that are updated in the nodes keep also the Max L dimension. In a invariant network we would always have scalars (distances, angles, etc) precomputed and the nodes would be updated only with those features, if I understood correctly.

**Reply by lucasdekam** [reply]
Yeah, the updatable features / hidden irreps $\mathrm{h}$ have the same max L as the messages; the update is linear. In an invariant network the inputs, messages and updatable states are all scalar (L=0). For example: https://pubs.rsc.org/en/content/articlehtml/2024/cp/d4cp01483a

**Reply by cmd23333** [reply]
@lucasdekam awesome explanation!
I have another question about how to match the equation in MACE paper with code.
This is the equation (8) in paper:
<img width="1126" height="104" alt="image" src="https://github.com/user-attachments/assets/29aff0bc-b2ae-4110-a374-834eefffed9c" />
(can't use markdown math here)
$$$
A^{(t)}_{i, kl_3m_3} = \sum_{l_1m_1, l_2m_2} C^{l_3m_3}_{l_1m_1, l_2m_2} \sum_{j \in N(i)} R^{(t)}_{kl_1l_2l_3}(r_{ji})Y^{m_1}_{l_1}(\hat{r}_{ji})\sum _{\widetilde{k}} W^{(t)}_{k\widetilde{k}l_2}h^{(t)}_{j,\widetilde{k}l_2m_2}
$$$

in code implementation, it use
```python
        node_feats = self.linear_up(node_feats)
        tp_weights = self.conv_tp_weights(edge_feats)
        mji = self.conv_tp(
            node_feats[sender], edge_attrs, tp_weights
        )  # [n_edges, irreps]
        message = scatter_sum(
            src=mji, index=receiver, dim=0, dim_size=num_nodes
        )  # [n_nodes, irreps]
```

we can find it in `class RealAgnosticInteractionBlock` or `class RealAgnosticResidualInteractionBlock`

and `node_feats = linear_up(...)` did the last sum(sum over W h in eq.), and we have already calculated `edge_attrs` (Y in eq.), and `tp_weights` calculate R in the eq.

and `mji` and `message` seems did the rest part, so my question is, where is the CG-coef in the equation ?

**Reply by ilyes319** [reply]
It is in the conv_tp class which is the e3nn tp or cueq if using cueq backend.
